flocker/node/test/test_loop.py

```diff
@@ -0,0 +1,583 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+Tests for ``flocker.node._loop``.
+"""
+
+from zope.interface import implementer
+
+from twisted.trial.unittest import SynchronousTestCase
+from twisted.test.proto_helpers import StringTransport, MemoryReactorClock
+from twisted.internet.protocol import Protocol, ReconnectingClientFactory
+from twisted.internet.defer import succeed, Deferred
+from twisted.protocols.amp import AMP
+
+from ...testtools import FakeAMPClient
+from .._loop import (
+    build_cluster_status_fsm, ClusterStatusInputs, _ClientStatusUpdate,
+    _StatusUpdate, _ClientConnected, ConvergenceLoopInputs,
+    ConvergenceLoopStates, build_convergence_loop_fsm, AgentLoopService,
+    ClusterStatus, ConvergenceLoop,
+    )
+from .._deploy import IDeployer, IStateChange
+from ...control._protocol import NodeStateCommand, _AgentLocator
+from ...control.test.test_protocol import iconvergence_agent_tests_factory
+
+
+def build_protocol():
+    """
+    :return: ``Protocol`` hooked up to transport.
+    """
+    p = Protocol()
+    p.makeConnection(StringTransport())
+    return p
+
+
+class StubFSM(object):
+    """
+    A finite state machine look-alike that just records inputs.
+    """
+    def __init__(self):
+        self.inputted = []
+
+    def receive(self, symbol):
+        self.inputted.append(symbol)
+
+
+class ClusterStatusFSMTests(SynchronousTestCase):
+    """
+    Tests for the cluster status FSM.
+    """
+    def setUp(self):
+        self.agent_operation = StubFSM()
+        self.fsm = build_cluster_status_fsm(self.agent_operation)
+
+    def assertConvergenceLoopInputted(self, expected):
+        """
+        Assert that that given set of symbols were input to the agent
+        operation FSM.
+        """
+        self.assertEqual(self.agent_operation.inputted, expected)
+
+    def test_creation_no_side_effects(self):
+        """
+        Creating the FSM has no side effects.
```

Comment:
Since it has no references to any other objects, talking to convergence loop is the only side-effect we care about. I mean yes it could do `os.remove` or whatever, but you can't test *everything*.